src/cpp/Cdr.cpp

```diff
@@ -2754,10 +2879,919 @@ Cdr& Cdr::deserializeWStringSequence(
+    assert(MEMBER_ID_INVALID == next_member_id_ || member_id == next_member_id_);
+    assert(current_state == Cdr::state(*this));
+    assert(EncodingAlgorithmFlag::PLAIN_CDR2 == current_encoding_ ||
+            EncodingAlgorithmFlag::DELIMIT_CDR2 == current_encoding_ ||
+            EncodingAlgorithmFlag::PL_CDR2 == current_encoding_);
+
+    if (is_present && EncodingAlgorithmFlag::PL_CDR2 == current_encoding_)
+    {
+        if (0x10000000 <= member_id.id)
+        {
+            throw BadParamException("Cannot serialize a member identifier equal or greater than 0x10000000");
+        }
+
+        switch (header_selection)
+        {
+            case XCdrHeaderSelection::SHORT_HEADER:
+            case XCdrHeaderSelection::AUTO_WITH_SHORT_HEADER_BY_DEFAULT:
+                xcdr2_serialize_short_member_header(member_id);
+                current_state.header_serialized_ = XCdrHeaderSelection::SHORT_HEADER;
+                break;
+            case XCdrHeaderSelection::LONG_HEADER:
+            case XCdrHeaderSelection::AUTO_WITH_LONG_HEADER_BY_DEFAULT:
+                xcdr2_serialize_long_member_header(member_id);
+                current_state.header_serialized_ = XCdrHeaderSelection::LONG_HEADER;
+                break;
+        }
+        current_state.header_selection_ = header_selection;
+    }
+
+    current_state.member_size_ = is_present ? 1 : 0;
+    current_state.next_member_id_ = member_id;
+    next_member_id_ = MEMBER_ID_INVALID;
+
+    return *this;
+}
+
+Cdr& Cdr::xcdr2_end_serialize_member(
+        const Cdr::state& current_state)
+{
+    assert(MEMBER_ID_INVALID != current_state.next_member_id_);
+    assert(EncodingAlgorithmFlag::PLAIN_CDR2 == current_encoding_ ||
+            EncodingAlgorithmFlag::DELIMIT_CDR2 == current_encoding_ ||
+            EncodingAlgorithmFlag::PL_CDR2 == current_encoding_);
+
+    if (0 < current_state.member_size_ && EncodingAlgorithmFlag::PL_CDR2 == current_encoding_)
+    {
+        auto last_offset = offset_;
+        set_state(current_state);
+        makeAlign(alignment(sizeof(uint32_t)));
+        if (NO_SERIALIZED_MEMBER_SIZE == serialized_member_size_)
+        {
+            const size_t member_serialized_size = last_offset - offset_ -
+                    (current_state.header_serialized_ == XCdrHeaderSelection::SHORT_HEADER ? 4 : 8);
+            if (8 < member_serialized_size)
+            {
+                switch (current_state.header_serialized_)
+                {
+                    case XCdrHeaderSelection::SHORT_HEADER:
+                        if (AUTO_WITH_SHORT_HEADER_BY_DEFAULT == current_state.header_selection_)
+                        {
+                            xcdr2_change_to_long_member_header(current_state.next_member_id_, member_serialized_size);
+                        }
+                        else
+                        {
+                            throw BadParamException("Cannot encode XCDRv2 LongMemberHeader");
```

Comment:
```suggestion
                            throw BadParamException("Cannot encode XCDRv2 ShortMemberHeader when serialized member size is greater than 8");
```

Reply:
We are trying to encode a long header, not the shorted one.